### PR TITLE
Refactor Critical Path code to prevent it from ever throwing exceptions

### LINKED ISSUE #1779: [Bug]: Prevent Critical Path code from throwing exceptions
### What happened?

Critical Path code occasionally throws Range exception here: https://github.com/jaegertracing/jaeger-ui/blob/17eb2916047f7c3d1f5691957e29bbf6784b2e22/packages/jaeger-ui/src/components/TracePage/CriticalPath/utils/sanitizeOverFlowingChildren.tsx#L87

### Steps to reproduce

Unknown, the failures have been happening sporadically in CI

### Expected behavior

Code should handle all input trace data, fall back gracefully to not returning CP if impossible to calculate.

### Relevant log output

_No response_

### Screenshot

_No response_

### Additional context

We need to rethink the [switch statement][1]. The relative positioning of parent/child spans has a fixed number of permutations, they should be covered by the appropriate if-conditions, ideally in a hierarchical manner, e.g. 
```
if child starts after parent
  if child ends before parent
    etc.
```

The function should also be simplified to do early returns. For example, instead of introducing a nesting level with
```
if (span && span.references.length) {
   // nested logic
}
```
replace it with
```
if (!(span && span.references.length)) {
   return
}
```

[1]: https://github.com/jaegertracing/jaeger-ui/blob/17eb2916047f7c3d1f5691957e29bbf6784b2e22/packages/jaeger-ui/src/components/TracePage/CriticalPath/utils/sanitizeOverFlowingChildren.tsx#L36

### Jaeger backend version

_No response_

### SDK

_No response_

### Pipeline

_No response_

### Stogage backend

_No response_

### Operating system

_No response_

### Deployment model

_No response_

### Deployment configs

_No response_

### PR DESCRIPTION
**Pull request by GLVSKiriti**
Resolves #1779 

In this PR switch statement is refactored to if-else block and covered all of our decision space.

### CODE REVIEW

File: packages/jaeger-ui/src/components/TracePage/CriticalPath/utils/sanitizeOverFlowingChildren.tsx [hunk 1]
```diff
@@ -27,80 +27,71 @@ const sanitizeOverFlowingChildren = (spanMap: Map<string, Span>): Map<string, Sp
...
-            // case 1: everything looks good
-            // |----parent----|
-            //   |----child--|
-            break;
+    if (!(span && span.references.length)) {
+      return;
+    }
+    // parentSpan will be undefined when its parentSpan is dropped previously
+    const parentSpan = spanMap.get(span.references[0].spanID);
+    if (parentSpan) {
```
Comment (yurishkuro) [inline packages/jaeger-ui/src/components/TracePage/CriticalPath/utils/sanitizeOverFlowingChildren.tsx:35]: reduce nesting further
```
if (!parentSpan) {
    spanMap.delete(span.spanID);
    return;
}
```

File: packages/jaeger-ui/src/components/TracePage/CriticalPath/utils/sanitizeOverFlowingChildren.tsx [hunk 2]
```diff
@@ -27,80 +27,71 @@ const sanitizeOverFlowingChildren = (spanMap: Map<string, Span>): Map<string, Sp
...
+          spanMap.delete(span.spanID);
 
-          case span.startTime < parentSpan.startTime &&
-            childEndTime <= parentEndTime &&
-            childEndTime > parentSpan.startTime:
-            // case 2: child start before parent, truncate is needed
+          // Remove the childSpanId from its parent span
+          parentSpan.childSpanIds = parentSpan.childSpanIds.filter(id => id === span.spanID);
+          spanMap.set(parentSpan.spanID, { ...parentSpan });
+        } else if (childEndTime > parentEndTime) {
```
Comment (yurishkuro) [inline packages/jaeger-ui/src/components/TracePage/CriticalPath/utils/sanitizeOverFlowingChildren.tsx:49]:
```suggestion
          spanMap.set(parentSpan.spanID, { ...parentSpan });
          return
        }
        if (childEndTime > parentEndTime) {
```

File: packages/jaeger-ui/src/components/TracePage/CriticalPath/utils/sanitizeOverFlowingChildren.tsx [hunk 3]
```diff
@@ -27,80 +27,71 @@ const sanitizeOverFlowingChildren = (spanMap: Map<string, Span>): Map<string, Sp
...
+          //                        |----child--|
+          // Remove the childSpan from spanMap
+          spanMap.delete(span.spanID);
 
-          case span.startTime < parentSpan.startTime &&
-            childEndTime <= parentEndTime &&
-            childEndTime > parentSpan.startTime:
-            // case 2: child start before parent, truncate is needed
+          // Remove the childSpanId from its parent span
+          parentSpan.childSpanIds = parentSpan.childSpanIds.filter(id => id === span.spanID);
```
Comment (yurishkuro) [inline packages/jaeger-ui/src/components/TracePage/CriticalPath/utils/sanitizeOverFlowingChildren.tsx:47]: `filter` function keeps the items if the condition is true. The result of this line will be `childSpanIds == [span.spanOD]`, is that what you wanted?

File: packages/jaeger-ui/src/components/TracePage/CriticalPath/utils/sanitizeOverFlowingChildren.tsx [hunk 4]
```diff
@@ -27,80 +27,71 @@ const sanitizeOverFlowingChildren = (spanMap: Map<string, Span>): Map<string, Sp
...
+        } else if (childEndTime > parentEndTime) {
+            // child end after parent, truncate is needed
             //      |----parent----|
-            //   |----child--|
+            //              |----child--|
             spanMap.set(span.spanID, {
               ...span,
-              startTime: parentSpan.startTime,
-              duration: childEndTime - parentSpan.startTime,
+              duration: parentEndTime - span.startTime,
```
Comment (yurishkuro) [inline packages/jaeger-ui/src/components/TracePage/CriticalPath/utils/sanitizeOverFlowingChildren.tsx:55]: this is a serious bug: you are overriding the source data of the trace, literally changing how it looks on the screen, not just decorating it. I think you can easily account for the difference in CP code by using some min/max functions.

File: packages/jaeger-ui/src/components/TracePage/CriticalPath/utils/sanitizeOverFlowingChildren.tsx [hunk 5]
```diff
@@ -27,80 +27,71 @@ const sanitizeOverFlowingChildren = (spanMap: Map<string, Span>): Map<string, Sp
...
             });
-            break;
+        }else{
+          // everything looks good
+          // |----parent----|
+          //   |----child--|
+        }
```
Comment (yurishkuro) [inline packages/jaeger-ui/src/components/TracePage/CriticalPath/utils/sanitizeOverFlowingChildren.tsx:61]:
```suggestion
          return;
        }
        // everything looks good
        // |----parent----|
        //   |----child--|
```

File: packages/jaeger-ui/src/components/TracePage/CriticalPath/utils/sanitizeOverFlowingChildren.tsx (hunk 3 above)
Comment (GLVSKiriti) [reply packages/jaeger-ui/src/components/TracePage/CriticalPath/utils/sanitizeOverFlowingChildren.tsx]: Yeah I think I did a typo here. I need result to be all child spanIds except span.spanID

Comment (yurishkuro) [reply packages/jaeger-ui/src/components/TracePage/CriticalPath/utils/sanitizeOverFlowingChildren.tsx]: so is this a bug? should it be caught by a unit test?

File: packages/jaeger-ui/src/components/TracePage/CriticalPath/utils/sanitizeOverFlowingChildren.tsx [hunk 6]
```diff
@@ -27,80 +27,76 @@ const sanitizeOverFlowingChildren = (spanMap: Map<string, Span>): Map<string, Sp
...
+        //      |----parent----|
+        //                        |----child--|
+        // Remove the childSpan from spanMap
+        spanMap.delete(span.spanID);
 
-            // Remove the childSpan from spanMap
-            spanMap.delete(span.spanID);
+        // Remove the childSpanId from its parent span
+        parentSpan.childSpanIds = parentSpan.childSpanIds.filter(id => id !== span.spanID);
+        spanMap.set(parentSpan.spanID, { ...parentSpan });
```
Comment (yurishkuro) [inline packages/jaeger-ui/src/components/TracePage/CriticalPath/utils/sanitizeOverFlowingChildren.tsx:53]: why is this needed? parentSpan is already in the map.

**Review by yurishkuro — APPROVED** [review]

File: packages/jaeger-ui/src/components/TracePage/CriticalPath/utils/sanitizeOverFlowingChildren.tsx [hunk 7]
```diff
@@ -27,80 +27,76 @@ const sanitizeOverFlowingChildren = (spanMap: Map<string, Span>): Map<string, Sp
...
-            throw RangeError(`Error while computing Critical Path Algorithm.`);
-        }
-      } else {
-        // Drop the child spans of dropped parent span
-        spanMap.delete(span.spanID);
-      }
+      // Remove the childSpanId from its parent span
+      parentSpan.childSpanIds = parentSpan.childSpanIds.filter(id => id !== span.spanID);
+      spanMap.set(parentSpan.spanID, { ...parentSpan });
+    } else if (childEndTime <= parentEndTime) {
```
Comment (yurishkuro) [inline packages/jaeger-ui/src/components/TracePage/CriticalPath/utils/sanitizeOverFlowingChildren.tsx:82]: codecov says this branch (child before parent) is not covered by unit tests, is that correct?

File: packages/jaeger-ui/src/components/TracePage/CriticalPath/utils/sanitizeOverFlowingChildren.tsx (hunk 6 above)
Comment (GLVSKiriti) [reply packages/jaeger-ui/src/components/TracePage/CriticalPath/utils/sanitizeOverFlowingChildren.tsx:53]: Here we are just updating the childSpanIds of parent span if any child span is removed.

Comment (yurishkuro) [reply packages/jaeger-ui/src/components/TracePage/CriticalPath/utils/sanitizeOverFlowingChildren.tsx:53]: map stores a reference to parentSpan object. You already updated its children in L52. What is the need to update the map (it should have no effect)?

File: packages/jaeger-ui/src/components/TracePage/CriticalPath/utils/sanitizeOverFlowingChildren.tsx (hunk 7 above)
Comment (GLVSKiriti) [reply packages/jaeger-ui/src/components/TracePage/CriticalPath/utils/sanitizeOverFlowingChildren.tsx:82]: Yeah, I will to add a new test case. Previously (child before parent )and (child after parent) are in same case block so that both cases are covered as there is already a test case written for (child after parent)

File: packages/jaeger-ui/src/components/TracePage/CriticalPath/utils/sanitizeOverFlowingChildren.tsx (hunk 4 above)
Comment (GLVSKiriti) [reply packages/jaeger-ui/src/components/TracePage/CriticalPath/utils/sanitizeOverFlowingChildren.tsx]: But we are just working on the copy of the spanMap which is received as parameter to the function sanitizeOverFlowingChildren function and We are also not modifying the original span data.

File: packages/jaeger-ui/src/components/TracePage/CriticalPath/utils/sanitizeOverFlowingChildren.tsx (hunk 7 above)
Comment (GLVSKiriti) [reply packages/jaeger-ui/src/components/TracePage/CriticalPath/utils/sanitizeOverFlowingChildren.tsx:80]: Added a new test case !!

File: packages/jaeger-ui/src/components/TracePage/CriticalPath/utils/sanitizeOverFlowingChildren.tsx (hunk 4 above)
Comment (GLVSKiriti) [reply packages/jaeger-ui/src/components/TracePage/CriticalPath/utils/sanitizeOverFlowingChildren.tsx]: Or instead can we send a copy of the trace data to the memoizedTraceCriticalPath function instead of sending data directly?? So that original data will not be effected

https://github.com/GLVSKiriti/jaeger-ui/blob/9da89cc3633442dabec033dbfec4eed2d8c74c30/packages/jaeger-ui/src/components/TracePage/index.tsx#L394

Here can we send memoizedTraceCriticalPath({...data}) ?? Will the problem solved then??

File: packages/jaeger-ui/src/components/TracePage/CriticalPath/utils/sanitizeOverFlowingChildren.test.js [hunk 8]
```diff
@@ -35,11 +36,12 @@ function getExpectedSanitizedData(spans, test) {
...
 
 describe.each([
-  [test3, new Map().set(test3.trace.spans[0].spanID, test3.trace.spans[0])],
-  [test4, new Map().set(test4.trace.spans[0].spanID, test4.trace.spans[0])],
+  [test3, new Map().set(test3.trace.spans[0].spanID, { ...test3.trace.spans[0], childSpanIds: [] })],
+  [test4, new Map().set(test4.trace.spans[0].spanID, { ...test4.trace.spans[0], childSpanIds: [] })],
   [test6, getExpectedSanitizedData(test6.trace.spans, 'test6')],
   [test7, getExpectedSanitizedData(test7.trace.spans, 'test7')],
   [test8, getExpectedSanitizedData(test8.trace.spans, 'test8')],
+  [test9, new Map().set(test9.trace.spans[0].spanID, { ...test9.trace.spans[0], childSpanIds: [] })],
```
Comment (yurishkuro) [inline packages/jaeger-ui/src/components/TracePage/CriticalPath/utils/sanitizeOverFlowingChildren.test.js:44]: doesn't need to be part of this PR, but why L31-34 call getExpectedSanitizedData in the test setup? What it means is the test is effectively asserting `f(x) == f(x)`, which does not test anything, only increases coverage, but not in a meaningful way.

**Comment by yurishkuro** [comment]
Thanks!

File: packages/jaeger-ui/src/components/TracePage/CriticalPath/utils/sanitizeOverFlowingChildren.tsx (hunk 4 above)
Comment (GLVSKiriti) [reply packages/jaeger-ui/src/components/TracePage/CriticalPath/utils/sanitizeOverFlowingChildren.tsx]: @yurishkuro can you clarify my doubt here!!

File: packages/jaeger-ui/src/components/TracePage/CriticalPath/utils/sanitizeOverFlowingChildren.test.js (hunk 8 above)
Comment (GLVSKiriti) [reply packages/jaeger-ui/src/components/TracePage/CriticalPath/utils/sanitizeOverFlowingChildren.test.js:44]: I don't get your point, can you elaborate it. test6,test7,test8 are needed to test the cases like when child overflows from parent(like child end after parent or start before parent etc..)

File: packages/jaeger-ui/src/components/TracePage/CriticalPath/utils/sanitizeOverFlowingChildren.tsx (hunk 4 above)
Comment (yurishkuro) [reply packages/jaeger-ui/src/components/TracePage/CriticalPath/utils/sanitizeOverFlowingChildren.tsx]: I think it's fine, I misread it as changing in place, but it's creating a clone of the span, so no issue.

File: packages/jaeger-ui/src/components/TracePage/CriticalPath/utils/sanitizeOverFlowingChildren.test.js (hunk 8 above)
Comment (yurishkuro) [reply packages/jaeger-ui/src/components/TracePage/CriticalPath/utils/sanitizeOverFlowingChildren.test.js:44]:
```
[test8, getExpectedSanitizedData(test8.trace.spans, 'test8')],
            ^^^^^^^^^^^^^^^^
```
I am referring to this part
